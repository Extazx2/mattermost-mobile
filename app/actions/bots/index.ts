// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Client4} from '@mm-redux/client';
import {BotTypes} from '@mm-redux/action_types';
import {bindClientFunc} from '@mm-redux/actions/helpers';

import {ActionFunc} from '@mm-redux/types/actions';

export function loadBot(botUserId: string): ActionFunc {
    return bindClientFunc({
        clientFunc: Client4.getBot,
        onSuccess: BotTypes.RECEIVED_BOT_ACCOUNT,
        params: [
            botUserId,
        ],
    });
}
