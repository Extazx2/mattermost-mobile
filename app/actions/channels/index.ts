export * from './action_objects';
export * from './requestor_actions';
export * from './high_order_actions';
export * from './channel_selection_actions';