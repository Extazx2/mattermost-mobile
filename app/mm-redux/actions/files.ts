// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Client4} from '@mm-redux/client';
import {FileTypes} from '@mm-redux/action_types';

import {Action, batchActions, DispatchFunc, GetStateFunc, ActionFunc} from '@mm-redux/types/actions';
import {FileUploadResponse, FileInfo} from '@mm-redux/types/files';
import {getFilesIdsForPost} from '@mm-redux/selectors/entities/files';

import {logError} from './errors';
import {bindClientFunc, forceLogoutIfNecessary} from './helpers';

export function getFilesForPost(postId: string): ActionFunc {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        let files;

        try {
            files = await Client4.getFileInfosForPost(postId);
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            dispatch(logError(error));
            return {error};
        }

        dispatch({
            type: FileTypes.RECEIVED_FILES_FOR_POST,
            data: files,
            postId,
        });

        return {data: files};
    };
}

export function getFilePublicLink(fileId: string): ActionFunc {
    return bindClientFunc({
        clientFunc: Client4.getFilePublicLink,
        onSuccess: FileTypes.RECEIVED_FILE_PUBLIC_LINK,
        params: [
            fileId,
        ],
    });
}

export function loadFilesForPostIfNecessary(postId: string): ActionFunc {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        const state = getState();
        let fileIds = getFilesIdsForPost(state, postId);

        if (!fileIds.length) {
            const {data: files} = await dispatch(getFilesForPost(postId));
            if (files) {
                fileIds = files.map((file: FileInfo) => file.id);
            }
        }

        return {data: fileIds};
    };
}
