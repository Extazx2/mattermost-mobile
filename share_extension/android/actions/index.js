// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getRedirectChannelForTeam} from '@mm-redux/selectors/entities/channels';

import {getMyChannelsAndMembersForTeam} from '@actions/channels';
import {ViewTypes} from 'app/constants';

export function getDefaultChannelForTeam(teamId) {
    return async (dispatch, getState) => {
        await dispatch(getMyChannelsAndMembersForTeam(teamId));

        const state = getState();
        const redirectChannel = getRedirectChannelForTeam(state, teamId);

        return redirectChannel;
    };
}

export function extensionSelectTeamId(teamId) {
    return {
        type: ViewTypes.EXTENSION_SELECTED_TEAM_ID,
        data: teamId,
    };
}
