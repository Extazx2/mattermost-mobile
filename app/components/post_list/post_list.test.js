// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import Preferences from '@mm-redux/constants/preferences';

import * as NavigationActions from 'app/actions/navigation';
import PostList from './post_list';

jest.useFakeTimers();
jest.mock('react-intl');

describe('PostList', () => {
    const serverURL = 'https://server-url.fake';
    const deeplinkRoot = 'mattermost://server-url.fake';

    const baseProps = {
        actions: {
            selectChannelFromDeepLinkMatch: jest.fn(),
            getChannelsByTeamName: jest.fn(),
            refreshChannelWithRetry: jest.fn(),
            selectFocusedPostId: jest.fn(),
            setDeepLinkURL: jest.fn(),
        },
        channelId: 'channel-id',
        deepLinkURL: '',
        lastPostIndex: -1,
        postIds: ['post-id-1', 'post-id-2'],
        serverURL,
        siteURL: 'https://site-url.fake',
        theme: Preferences.THEMES.default,
    };

    const deepLinks = {
        permalink: deeplinkRoot + '/team-name/pl/pl-id',
        channel: deeplinkRoot + '/team-name/channels/channel-name',
    };

    test('should match snapshot', () => {
        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('setting permalink deep link', () => {
        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );
        const instance = wrapper.instance();
        const handlePermalinkPress = jest.spyOn(instance, 'handlePermalinkPress')

        wrapper.setProps({deepLinkURL: deepLinks.permalink});
        expect(baseProps.actions.setDeepLinkURL).toHaveBeenCalled();
        expect(handlePermalinkPress).toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('handlePermalinkPress calls props.onPermalinkPress set', async () => {
        const props = {
            ...baseProps,
            onPermalinkPress: jest.fn(),
        };
        const wrapper = shallow(
            <PostList {...props}/>,
        );
        const instance = wrapper.instance();

        await instance.handlePermalinkPress();
        expect(props.onPermalinkPress).toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('handlePermalinkPress calls permalinkBadTeam when props.onPermalinkPress is not set and getChannelsByTeamName fails', async () => {
        const props = {
            ...baseProps,
            onPermalinkPress: null,
        };
        props.actions.getChannelsByTeamName.mockResolvedValue({error: true});

        const wrapper = shallow(
            <PostList {...props}/>,
        );
        const instance = wrapper.instance();
        instance.permalinkBadTeam = jest.fn();

        await instance.handlePermalinkPress();
        expect(instance.permalinkBadTeam).toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('handlePermalinkPress calls showPermalinkView when props.onPermalinkPress is not set and getChannelsByTeamName succeeds', async () => {
        const props = {
            ...baseProps,
            onPermalinkPress: null,
        };
        props.actions.getChannelsByTeamName.mockResolvedValue({data: true});

        const wrapper = shallow(
            <PostList {...props}/>,
        );
        const instance = wrapper.instance();
        instance.showPermalinkView = jest.fn();

        await instance.handlePermalinkPress();
        expect(instance.showPermalinkView).toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('showPermalinkView calls actions.selectFocuesPostId but doesn\'t open modal if permalink already showing', () => {
        const showModalOverCurrentContext = jest.spyOn(NavigationActions, 'showModalOverCurrentContext');
        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );
        const instance = wrapper.instance();
        instance.showingPermalink = true;

        instance.showPermalinkView();
        expect(baseProps.actions.selectFocusedPostId).toHaveBeenCalled();
        expect(showModalOverCurrentContext).not.toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('showPermalinkView calls actions.selectFocuesPostId and opens modal if permalink is not showing', () => {
        const showModalOverCurrentContext = jest.spyOn(NavigationActions, 'showModalOverCurrentContext');
        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );
        const instance = wrapper.instance();
        instance.showingPermalink = false;

        instance.showPermalinkView();
        expect(baseProps.actions.selectFocusedPostId).toHaveBeenCalled();
        expect(showModalOverCurrentContext).toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('setting channel deep link', () => {
        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );

        wrapper.setProps({deepLinkURL: deepLinks.channel});
        expect(baseProps.actions.setDeepLinkURL).toHaveBeenCalled();
        expect(baseProps.actions.selectChannelFromDeepLinkMatch).toHaveBeenCalled();
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should call flatListScrollToIndex only when ref is set and index is in range', () => {
        jest.spyOn(global, 'requestAnimationFrame').mockImplementation((cb) => cb());

        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );
        const instance = wrapper.instance();
        const flatListScrollToIndex = jest.spyOn(instance, 'flatListScrollToIndex');
        const indexInRange = baseProps.postIds.length;
        const indexOutOfRange = [-1, indexInRange + 1];

        instance.flatListRef = {
            current: null,
        };
        instance.scrollToIndex(indexInRange);
        expect(flatListScrollToIndex).not.toHaveBeenCalled();

        instance.flatListRef = {
            current: {
                scrollToIndex: jest.fn(),
            },
        };
        for (const index of indexOutOfRange) {
            instance.scrollToIndex(index);
            expect(flatListScrollToIndex).not.toHaveBeenCalled();
        }

        instance.scrollToIndex(indexInRange);
        expect(flatListScrollToIndex).toHaveBeenCalled();
    });

    test('should load more posts if available space on the screen', () => {
        const wrapper = shallow(
            <PostList {...baseProps}/>,
        );
        const instance = wrapper.instance();
        instance.loadToFillContent = jest.fn();

        wrapper.setProps({
            extraData: false,
        });
        expect(instance.loadToFillContent).toHaveBeenCalledTimes(0);

        instance.postListHeight = 500;
        instance.contentHeight = 200;
        wrapper.setProps({
            extraData: true,
        });

        expect(instance.loadToFillContent).toHaveBeenCalledTimes(1);
    });
});
