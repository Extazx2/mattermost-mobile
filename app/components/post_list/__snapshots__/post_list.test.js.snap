// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList handlePermalinkPress calls permalinkBadTeam when props.onPermalinkPress is not set and getChannelsByTeamName fails 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList handlePermalinkPress calls props.onPermalinkPress set 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList handlePermalinkPress calls showPermalinkView when props.onPermalinkPress is not set and getChannelsByTeamName succeeds 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList setting channel deep link 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList setting permalink deep link 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList should match snapshot 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList showPermalinkView calls actions.selectFocuesPostId and opens modal if permalink is not showing 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;

exports[`PostList showPermalinkView calls actions.selectFocuesPostId but doesn't open modal if permalink already showing 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingTop": 5,
    }
  }
  data={
    Array [
      "post-id-1",
      "post-id-2",
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      "channel-id",
      undefined,
      undefined,
      undefined,
    ]
  }
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  keyboardDismissMode="interactive"
  keyboardShouldPersistTaps="handled"
  listKey="recyclerFor-channel-id"
  maintainVisibleContentPosition={
    Object {
      "autoscrollToTopThreshold": 60,
      "minIndexForVisible": 0,
    }
  }
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollToIndexFailed={[Function]}
  refreshControl={
    <RefreshControl
      colors={
        Array [
          "#3d3c40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
      tintColor="#3d3c40"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={60}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={50}
/>
`;
